address_book/src/lib.rs

```diff
@@ -32,31 +31,101 @@ static LOGGING: Lazy<()> = Lazy::new(|| unsafe {
     tide::log::with_level(LOG_LEVEL);
 });
 
+// TODO: remove test store after tests, add TestFileStore type that implements Drop?
+
+trait Store {
+    fn save(&self, address: &UserAddress, pubkey: &UserPubKey) -> Result<(), std::io::Error>;
+    fn load(&self, address: &UserAddress) -> Option<UserPubKey>;
+}
+
+#[derive(Debug, Clone)]
+struct FileStore {
+    dir: PathBuf,
+}
+
+impl FileStore {
+    fn new(dir: PathBuf) -> Self {
+        Self { dir }
+    }
+
+    fn path(&self, address: &UserAddress) -> PathBuf {
+        let as_hex = hex::encode(bincode::serialize(&address).unwrap());
+        self.dir.join(format!("{}.bin", as_hex))
+    }
+
+    fn tmp_path(&self, address: &UserAddress) -> PathBuf {
+        let rand_string: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(16)
+            .map(char::from)
+            .collect();
+
+        self.path(address).with_extension(rand_string)
+    }
+}
+
+impl Store for FileStore {
+    fn save(&self, address: &UserAddress, pubkey: &UserPubKey) -> Result<(), std::io::Error> {
+        let tmp_path = self.tmp_path(address);
+        fs::write(&tmp_path, bincode::serialize(&pubkey).unwrap())?;
+        fs::rename(&tmp_path, self.path(address))
+    }
+
+    fn load(&self, address: &UserAddress) -> Option<UserPubKey> {
+        match fs::read(self.path(address)) {
+            Ok(bytes) => Some(bincode::deserialize(&bytes).unwrap()),
+            Err(_) => None,
+        }
+    }
+}
+
 #[derive(Debug, Deserialize, Serialize)]
 pub struct InsertPubKey {
     pub pub_key_bytes: Vec<u8>,
     pub sig: Signature,
 }
 
-#[derive(Clone, Default)]
+#[derive(Clone)]
 struct ServerState {
-    map: Arc<RwLock<HashMap<UserAddress, UserPubKey>>>,
+    store: FileStore,
 }
 
 pub fn address_book_port() -> String {
     std::env::var("PORT").unwrap_or_else(|_| DEFAULT_PORT.to_string())
 }
 
+pub fn cape_data_path() -> PathBuf {
+    dirs::data_local_dir()
```

Comment:
@nyospe Took this from PR #504 (with a bit of rewriting).

Reply:
👍